From 63f72729f39d85003863090a269542a046145c34 Mon Sep 17 00:00:00 2001
From: Sergey Sharybin <sergey@blender.org>
Date: Thu, 27 Mar 2025 15:52:26 +0100
Subject: [PATCH] Fix baked kernels failure when application has space in its
 path

When baked kernels are used and the application contains space in the
part to its executable rendering will fail on Linux.

The exact details of why it happens are not very clear from our side,
but seemingly it is caused by the fact that HIP-RT is passing a large
global array to a different library (HIP driver, via hipModuleLoadData).
Having global variables of such size seems to be always problematic as
they can not be stored on stack and, possibly, extra mapping is involved
here. It is not clear whether it is a quirk of the HIP driver, or Linux,
or, maybe, something completely different.

It is possible to work-around the problem by making a temporary
copy of data on heap memory and pass it to the hipModuleLoadData().
This is how other areas are dealing with modules in Blender.

This change contains patch against HIP-RT and the new HIP-RT
library compiled with the patch. It seems to fix the problem
reported in the report.

Original report on Blender side: https://projects.blender.org/blender/blender/issues/134992
Original PR on Blender side: https://projects.blender.org/blender/blender/pulls/135403
---
 contrib/Orochi/ParallelPrimitives/RadixSort.cpp | 12 +++++++++++-
 hiprt/impl/Compiler.cpp                         | 10 +++++++++-
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/contrib/Orochi/ParallelPrimitives/RadixSort.cpp b/contrib/Orochi/ParallelPrimitives/RadixSort.cpp
index 8a075fa..8b1700e 100644
--- a/contrib/Orochi/ParallelPrimitives/RadixSort.cpp
+++ b/contrib/Orochi/ParallelPrimitives/RadixSort.cpp
@@ -26,6 +26,7 @@
 #include <cassert>
 #include <iostream>
 #include <numeric>
+#include <vector>
 
 // if ORO_PP_LOAD_FROM_STRING &&     ORO_PRECOMPILED -> we load the precompiled/baked kernels.
 // if ORO_PP_LOAD_FROM_STRING && NOT ORO_PRECOMPILED -> we load the baked source code kernels (from Kernels.h / KernelArgs.h)
@@ -242,7 +243,16 @@ void RadixSort::compileKernels( const std::string& kernelPath, const std::string
 	{
 		if constexpr( usePrecompiledAndBakedKernel )
 		{
-			oroFunctions[record.kernelType] = m_oroutils.getFunctionFromPrecompiledBinary_asData(oro_compiled_kernels_h, oro_compiled_kernels_h_size, record.kernelName.c_str() );
+			// Copy the data into a heap memory.
+			//
+			// Otherwise it seems to be causing issues with access from the HIP SDK when the application binary is
+			// located in a directory with space in it.
+			//
+			// The speculation is that static global memory can not really be megabytes, and access to it requires
+			// mapping of the original file, and this is where things start to go wrong.
+			std::vector<unsigned char> binary(oro_compiled_kernels_h, oro_compiled_kernels_h + oro_compiled_kernels_h_size);
+
+			oroFunctions[record.kernelType] = m_oroutils.getFunctionFromPrecompiledBinary_asData(binary.data(), binary.size(), record.kernelName.c_str() );
 		}
 		else if constexpr( useBakeKernel )
 		{
diff --git a/hiprt/impl/Compiler.cpp b/hiprt/impl/Compiler.cpp
index 514667a..509f3f4 100644
--- a/hiprt/impl/Compiler.cpp
+++ b/hiprt/impl/Compiler.cpp
@@ -776,7 +776,15 @@ oroFunction Compiler::getFunctionFromPrecompiledBinary( const std::string& funcN
 			// kernel.
 			if constexpr ( UseBakedCompiledKernel )
 			{
-				checkOro( oroModuleLoadData( &module, &bvh_build_array_h ) );
+				// Copy the data into a heap memory.
+				//
+				// Otherwise it seems to be causing issues with access from the HIP SDK when the application binary is
+				// located in a directory with space in it.
+				//
+				// The speculation is that static global memory can not really be megabytes, and access to it requires
+				// mapping of the original file, and this is where things start to go wrong.
+				std::vector<char> binary(bvh_build_array_h, bvh_build_array_h + bvh_build_array_h_size);
+				checkOro( oroModuleLoadData( &module, binary.data() ) );
 			}
 			else
 			{
-- 
2.48.1

